### DISCUSSION TITLE
Studying Singularities in the iCub Arm's Workspace

### DISCUSSION THREAD

**Discussion by Woolfrey** [robotology/community#559] [General]
### Summary:
-	The iCub arm has an ill-conditioned Jacobian for most of its possible arm configurations compared to other robots. This makes classical inverse kinematics method difficult to implement.
-	The region of the workspace used for manipulating objects in front of the robot has low manipulability.
-	This analysis has not included the torso, which could improve results.

### Background:
I am part of the ergoCub project and have been using the iCub to develop a Cartesian controller for manipulating objects. The velocity of the hand is related to the joint velocities by the Jacobian matrix:

![image](https://user-images.githubusercontent.com/62581255/138671520-17d799d6-e247-4266-b7c8-cd7cdd21c25d.png)

A general solution for the joint velocities to control the speed of the hand takes the form:

![image](https://user-images.githubusercontent.com/62581255/138671661-6ff461ab-d519-436a-944e-efb9a2a7839b.png)

where the pseudoinverse is:

![image](https://user-images.githubusercontent.com/62581255/138671739-c1d1f946-c2e4-45f5-a664-6f0a9d253520.png)

Two problems can occur:

1. If the Jacobian is singular, the inverse Jacobian does not exist.
2. Near singularities, joint velocities grow to infinity and the robot can lose control.

A solution is to use damped least squares when near singularities. This causes task error, but avoids instability. It can be activated at a threshold value to preserve accuracy where possible. I found that the iCub was often near-singular and would rarely execute a hand motion accurately.

I decided to investigate.

### Measuring Proximity to a Singularity:

The manipulability index is given by:

![image](https://user-images.githubusercontent.com/62581255/138671961-1f513d4f-a306-4a75-92a2-eb74848f38ff.png)

which is the product of singular values of the Jacobian. Hence, at a singularity, mu = 0. In my experience, mu > 0.01 is quite good.

I reconstructed a kinematic model of the iCub arm (v2) in MATLAB based off the DH parameters (excluding the torso links). For each joint, I iterated through 6 equispaced positions between limits, for a total of 6^7 = 279,963 possible joint configurations. I computed the manipulability indexf or each.

_NOTE: Self-collisions are not accounted for, so not all configurations are feasible._

I compared these values to other commercially available robot arms:

- The UR3 by Universal Robot (6DOF, reach of 500mm),
- The LBR iiwa 7 by Kuka (7DOF, reach of 800mm),
- The robotlink by igus (6DOF, reach of 600mm).

Here some basic statistics on the manipulability for each of the arms:

| 		| Maximum | Median* |
|--------------------|--------------|--------------|
| iCub		| 0.0152     | 0.0036      |
| UR3 		| 0.0162     | 0.0028      |
| LBR iiwa	| 0.1477     | 0.0597      |
| robolink	| 0.0217     | 0.0016      |

*only non-zero values were used to compute the median.

And here is a histogram showing the distribution of values:

![image](https://user-images.githubusercontent.com/62581255/138671030-f367f84b-d708-43af-9744-f5018e21ad7c.png)

As can be seen, the distribution for the iCub arm is left-skewed. The Jacobian is almost always ill-conditioned. This makes it very difficult to apply classical inverse kinematics methods for controlling the hand. This could be an issue for Cartesian impedance control, since the Cartesian inertia matrix is also computed with the Jacobian:

![image](https://user-images.githubusercontent.com/62581255/138672435-a13e8dd1-924b-497d-ad3b-83a87754d5d2.png)

### Examining the Workspace:

I computed the hand position of the iCub arm for every configuration. I then generated a point cloud for each hand position, using the manipulability index to colour the points: green is good, red is bad (MATLAB wouldn't let me plot a model of the arm and use colour scales simultaneously for some reason):

![image](https://user-images.githubusercontent.com/62581255/138670548-e63ac329-39dc-4f09-bfe6-11099fd9ecfe.png)

![image](https://user-images.githubusercontent.com/62581255/138670862-c67db3f8-5c8e-4f29-844a-18dffd9468ae.png)

The general working position for the iCub when manipulating objects is down and toward the front. This region has poor manipulability. Ironically, the best manipulability appears to be up near the head. Of course, the workspace of the arm can be moved by moving the torso.

### Conclusions:
-	The Jacobian for the iCub arm is ill-conditioned for most of its possible configurations compared to other light-weight, collaborative robot arms. 
-	There are very few poses with “good” manipulability.
-	Care must be taken when using classical control methods to ensure good manipulability of the arm.
-	The torso joints should probably be used to increase the range of the workspace and find better configurations for manipulability.
-	The ipopt algorithm within the iCartesianControl is quite robust – it would be interesting to know why!

**Comment by pattacini** [comment]
Nice writeup @Woolfrey 👍🏻 

My educated guess is that, compared with the other more industrial manipulators, the iCub arm is kind of penalized in terms of both size and range of motion for some of the joints and specifically for the shoulder roll and the wrist yaw, especially when considering the area in front of the robot where objects are mostly supposed to lie for manipulation purposes.

However, as you said, if we enable the torso joints the story changes quite considerably.

Here's below the manipulability for the 10-DOF chain [[code](https://gist.github.com/pattacini/b8e6fc00c4de4dd6ff8e30c7dddb26df), [data](https://github.com/robotology/community/files/7413615/manipulability.txt)].

![manipulability](https://user-images.githubusercontent.com/3738070/138773096-2a6b2178-feff-42d2-a795-e746330a3b85.png)

As it's visible, numbers are much better now.
This is also in line with my past experience with the robot: no matter what method for IK is employed, without the torso, there's very little chance of success.

Ipopt doesn't deal directly with the inversion of the kinematic Jacobian, being a state-of-the-art interior-point primal-dual method for nonlinear constraint optimization. This is just to give a shallow hint on the fact that Ipopt tends to outperform traditional IK methods also in the 10-DOF configuration (see this [resource](https://github.com/robotology/community/discussions/208#discussioncomment-163860)), right because Jacobian-based approaches can get trapped bumping into joint bounds when possible paths to the solution may be found only by exploring directions not proposed by the Jacobian per se.

**Comment by Woolfrey** [comment]
### Summary

- The iCub 3 arm has better manipulability than the iCub 2. This means it should be able to perform more dexterous tasks than the iCub 2.
- Including the torso dramatically increases the workspace and manipulability. It is critical to make use of the torso if one wishes to avoid singularities, but also exploit redundancy.
- Utilizing the torso for two-handed manipulation is not as simple as both arms must share the torso joints. Further analysis is required for this scenario.

I decided to do a further analysis comparing the iCub 2 and iCub 3, and then the iCub 3 with and without the torso.

Below are some basic statistics summarising the manipulability between them:

| Model | Maximum | Median |
|--------|------------|----------|
| iCub 2 (No torso) | 0.0152 | 0.0036 |
| iCub 3 (No torso) | 0.0136 | 0.0040 |
| iCub 3 (With torso) | 0.1621 | 0.0783 |

Interestingly, the iCub 2 has a better maximum manipulability than the iCub 3. Although, this could be because discretized joint combinations are used so the "best" configuration might not have been captured for the iCub 3. Confirming @pattacini's previous post, the inclusion the torso joint dramatically increases the manipulability. 

### iCub 2 vs iCub 3 (Left Arm)

Below is a histogram of the manipulability for both the robots using 6 equispaced joint positions between limits (6^7 = 279, 936 combinations). The distribution for v3 is less skewed, resulting in a higher median value. This means the iCub 3 has more joint configurations with better manipulability. If I am not mistaken, this is because the iCub 3 has slightly longer limbs which means a smaller change in the joint angles is required to produce the same motion at the hand.

![v2v3histogram](https://user-images.githubusercontent.com/62581255/139835014-3699a391-f2c1-49c2-9424-ef9f60b549d5.jpg)

Here is a point cloud comparing the workspaces, coloured by manipulability (hopefully I got the dimensions correct). The iCub 3 has a larger workspace. More interestingly, the regions of greatest manipulability are in front of the robot - where one might expect to be doing the majority of manipulation tasks.

![v2v3scatter](https://user-images.githubusercontent.com/62581255/139835595-2ec288de-5ae6-48c9-8573-9fa8555b1c2e.jpg)

### Including the Torso

The torso adds 3 additional joints to the kinematic chain. For this analysis, I used 4 equispaced joint positions between the upper and lower joint limits. I excluded the first joint to reduce the computational cost since it can be proved that:

![image](https://user-images.githubusercontent.com/62581255/139836983-1c68a351-a60e-4d7d-942b-27a4562159b5.png)

In total there were 4^9 = 262, 144 combinations.

Below is a histogram comparing the range of manipulability for the iCub 3 arm with and without the torso joints:

![icub3histogram](https://user-images.githubusercontent.com/62581255/139837120-a3d16d5c-5306-4bce-a73a-587558fca942.jpg)

There is a much more even distribution, but also many more joint combinations with better manipulability. This means it is easier to find joint configurations that are non-singular, making it easier to solve the inverse kinematics. Moreover, there are more feasible solutions within the joint limits.

Here is a point cloud of the workspace when including the torso:

![image](https://user-images.githubusercontent.com/62581255/139838022-684b3e43-2db0-4e0f-9273-7290c9f6cf7d.png)

The region with better manipulability (green) is much larger than before. Moreover, it appears to cover the area to the front of the robot which would be the ideal workspace for object manipulation. Note that since the volume of the workspace could be dramatically increased by including the first joint (though this would require over 1 million joint configurations to be computed).

### Conclusion

As @pattacini previously confirmed, using the torso on the iCub is essential for using inverse kinematics. It not only increases the workspace, but also the manipulability. It also provides 3DOF of redundancy that can be utilized for optimising the joint configuration (joint limit avoidance, singularity avoidance, energy minimization, etc.).

What remains to be investigated is the two-handed manipulation. The extra range of motion afforded to a single arm  is negated if both hands need to maintain good manipulability. This might be worth analysing in the future.

**Reply by pattacini** [reply]
>What remains to be investigated is the two-handed manipulation. The extra range of motion afforded to a single arm is negated if both hands need to maintain good manipulability. This might be worth analysing in the future.

For this, you may reach out to the [AMI](https://www.iit.it/web/ami) group for their work on lifting a weight with 2 hands. (cc @DanielePucci @lrapetti)

Also, find below a couple of works we did in the past to address the IK of a bimanual task:
- 📄 "Automatic kinematic chain calibration using artificial skin: Self-touch in the iCub humanoid robot"[^1] (cc @alecive @matejhof)
- 📄 "A novel pipeline for bi-manual handover task"[^2] (cc @giuliavezzani @lornat75)

[^1]: @article{10.1109/icra.2014.6907178, 
year = {2014}, 
title = {{Automatic kinematic chain calibration using artificial skin: Self-touch in the iCub humanoid robot}}, 
author = {Roncone, Alessandro and Hoffmann, Matej and Pattacini, Ugo and Metta, Giorgio}, 
doi = {10.1109/icra.2014.6907178}, 
pages = {2305--2312}, 
keywords = {}
}

[^2]: @article{10.1080/01691864.2017.1380535, 
year = {2017}, 
title = {{A novel pipeline for bi-manual handover task}}, 
author = {Vezzani, G and Regoli, M and Pattacini, U and Natale, L}, 
journal = {Advanced Robotics}, 
issn = {0169-1864}, 
doi = {10.1080/01691864.2017.1380535}, 
pages = {1--14}, 
keywords = {}
}
